FLT/NumberField/AdeleRing.lean

```diff
@@ -18,6 +23,141 @@ end LocallyCompact
 
 section BaseChange
 
+namespace NumberField.AdeleRing
+
+variable (K L : Type*) [Field K] [Field L] [NumberField K] [NumberField L] [Algebra K L]
+
+noncomputable instance : Algebra K (NumberField.AdeleRing L) :=
+  RingHom.toAlgebra <| (algebraMap L (NumberField.AdeleRing L)).comp <| algebraMap K L
+
+noncomputable abbrev tensorProductLinearEquivPi : AdeleRing K ⊗[K] L ≃ₗ[K]
+    (Fin (Module.finrank K L) → AdeleRing K) :=
+  LinearEquiv.trans (TensorProduct.congr (LinearEquiv.refl K (AdeleRing K))
+      (Basis.equivFun (Module.finBasis K L)))
+    (TensorProduct.piScalarRight _ _ _ _)
+
+variable {L}
+
+theorem tensorProductLinearEquivPi_tsum_apply (l : L) :
+    tensorProductLinearEquivPi K L (1 ⊗ₜ[K] l) =
+      fun i => algebraMap _ _ (Module.finBasis K L |>.equivFun l i) := by
+  simp [tensorProductLinearEquivPi, Algebra.algebraMap_eq_smul_one]
+
+variable {K}
+
+theorem tensorProductLinearEquivPi_symm_apply_of_algebraMap
+    (x : Fin (Module.finrank K L) → K) :
+    (tensorProductLinearEquivPi K L).symm (fun i => algebraMap _ _ (x i)) =
+      1 ⊗ₜ[K] ((Module.finBasis K L).equivFun.symm x) := by
+  simp only [LinearEquiv.trans_symm, LinearEquiv.trans_apply,
+    Algebra.TensorProduct.piScalarRight_symm_apply_of_algebraMap, TensorProduct.congr_symm_tmul,
+    LinearEquiv.refl_symm, LinearEquiv.refl_apply, map_sum, Basis.equivFun_symm_apply]
+  rw [Finset.sum_comm]
+  simp only [← Finset.sum_smul, Fintype.sum_pi_single]
+
+/-
+smmercuri: A possible alternative using `moduleTopology` is
+```
+instance :
+    TopologicalSpace (NumberField.AdeleRing K ⊗[K] L) :=
+  letI := TopologicalSpace.induced (algebraMap K (AdeleRing K)) inferInstance
+  moduleTopology K _
+```
+However it is not clear to me how the inverse function is of `tensorProductLinearEquivPi`
+is continuous in that case. Additionally,
+https://math.mit.edu/classes/18.785/2017fa/LectureNotes25.pdf (just above Prop 25.10)
+for an informal source where the tensor product is given the product topology. Maybe they
+coincide!
+-/
```

Comment:
Yes, the module topology on a finite free R-module coincides with the product topology for any choice of basis (and in particular the topology on a finite free thing is independent of choice of basis). This should now all be in mathlib.

Reply:
Note to self; I haven't read the rest of this file.